### PR TITLE
Feature/ntgr 48/move locallib functions to class directory

### PR DESCRIPTION
**Pull request by naveenbalachandar**
## Description of the change

Moved the following functions in locallib.php to a new classes/local/credentials.php

- `accredible_get_credentials` -> `credentials->get_credentials`
- `create_credential` -> `credentials->create_credential`
- `create_credential_legacy` -> `credentials->create_credential_legacy`
- `accredible_check_for_existing_credential` -> `credentials->check_for_existing_credential`
- `accredible_check_for_existing_certificate` -> `credentials->check_for_existing_certificate`

All coding style standards from the code-checker plugin has been followed for the new additions.

## Testing
- Added unit tests for all the above functions in tests/classes/local/credentials_test.php
- Added unit tests for get_credentials, create_credentials api calls in tests/classes/apiRest/apiRest_test.php

## Manual testing

**credentials->get_credentials**
- Click on any accredible activity with existing credential
- The view page will call this function to fetch all credentials of the group

**credentials->create_credential**
- Add an accredible activity and manually trigger a certificate issue.
- This function will be called to create a new credential using the public accredible API

**credentials->create_credential_legacy** (legacy)
- change the if statement in lib.php#128 to always fail and update the database table -> mdl_accredible to have a achievement_id (group_name)
- Edit an existing accredible activity and add a new user under the "Manually issue certificates/badges" and save

**credentials->check_for_existing_credential**
- Add a final quiz to the course and trigger it for course completion.
- Login as a student and attempt the quiz and submit
- The accredible_quiz_submission_handler will call this function to check if a credential exists for a user

**credentials->check_for_existing_certificate** (legacy)
- Edit line 357 in locallib.php and trigger the else statement
- repeat the steps for credentials->check_for_existing_credential

### CODE REVIEW

**Review by kei178 — CHANGES_REQUESTED** [review]
I haven't done manual testing yet. I left three comments. Aside from that it all looks good.

File: tests/classes/local/credentials_test.php [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+            ->method('get')
+            ->with($this->equalTo($url))
+            ->willReturn($resdata);
+
+        // Expect to return an empty array.
+        $api = new apiRest($mockclient3);
+        $localcredentials = new credentials($api);
+        $result = $localcredentials->get_credentials(9549);
+        $this->assertEquals($result, array());
+    }
```
[the remark this answers is not in the document]
Comment (kei178) [reply tests/classes/local/credentials_test.php:284]: The following example would be more suitable than `willReturnOnConsecutiveCalls` for the pagination test case.

```php
$mock->expects($this->at(0))
    ->method('foo')
    ->with(...)
    ->will($this->returnValue(...));

$mock->expects($this->at(1))
    ->method('foo')
    ->with(...)
    ->will($this->returnValue(...));
```

File: classes/local/credentials.php [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            // Log an event now we've created the credential if possible.
+            if ($event != null) {
+                $certificateevent = \mod_accredible\event\certificate_created::create(array(
+                    'objectid' => $credential->credential->id,
+                    'context' => context_module::instance($event->contextinstanceid),
+                    'relateduserid' => $event->relateduserid,
+                    'issued_on' => $issuedon
+                ));
+                $certificateevent->trigger();
+            }
```
[the remark this answers is not in the document]
Comment (naveenbalachandar) [reply classes/local/credentials.php]: Makes sense. I have removed this logic as it is not being used and removed the argument.

File: tests/classes/local/credentials_test.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (naveenbalachandar) [reply tests/classes/local/credentials_test.php:288]: `$this->at()` is deprecated and will be removed from PHPUnit 10.
So I have used withConsecutive and onConsecutiveCalls for this.

**Review by kei178 — APPROVED** [review]
LGTM.

**Comment by kei178** [comment]
This branch will be merged into `develop` after the current QA process is completed.

**Pull request merged**